[#117] Send Sentry report if file path given for logs is not valid

ssawchenko opened the pull request:
### Related Issue: #117 


### Summary of Problem:
We are seeing issues where some of the standard file paths are not actually valid on some devices - and as such writes to these file locations are failing. We would like Steamclog to report when this is occurring for it's log files.

### Proposed Solution:
* If the file path given is null or does not exist, fire an error report to Sentry
* To avoid possible noisy reports for this issue, I have tried to use a DataStore to store a value that lets us know if we've already logged this issue.
* Sample app allows us to toggle between valid and invalid file paths

### Testing Steps:


### Screenshots:

ssawchenko commented on steamclog/src/main/java/com/steamclock/steamclog/Steamclog.kt:
```diff
@@ -48,13 +51,41 @@ object SteamcLog {
+                    logInternal(LogLevel.Warn, sentryErrorTitle)
+                }
+            } catch (e: Exception) {
+                logInternal(LogLevel.Info, "Could not read local DataStore")
```
Safeguarding against any error that might occur with DataStore; since this is the first I've worked with it I want to make sure that failures with it do cause the app to crash.

phileo commented on app/src/main/java/com/steamclock/steamclogsample/MainActivity.kt:
```diff
@@ -106,13 +125,32 @@ class MainActivity : AppCompatActivity() {
+                launch {
+                    // Doesn't need to be lifecycle aware for this test
+                    val timestamp =
+                    AppDataStore(baseContext).apply {
```
curious to know why `baseContext` is passed in here here, but `applicationContext` is passed into `SClogDataStore`

ssawchenko replied:
Hrm. Honestly I think that was a discrepancy I missed. Because we're using the context to lookup the datastore instance I don't *think* it matters which we use (although I could be wrong). Since some of the examples I have found seem to make use of the application context, I'll change these cases to explicitly request the `Application` to remove ambiguity.

phileo replied:
passing in Application is the right call.  Activity Context is tied to the Activity lifecycle, so we don't want any potential issues from carrying out operations tied to an Activity lifecycle.

ssawchenko commented on steamclog/src/main/java/com/steamclock/steamclog/Steamclog.kt:
```diff
@@ -48,13 +51,41 @@ object SteamcLog {
+                // call the datastore methods; we need to use this with caution, but
+                // https://developer.android.com/topic/libraries/architecture/datastore#synchronous
+                val dataStore = SClogDataStore(context)
+                val alreadyReportedFailure = runBlocking { dataStore.getHasReportedFilepathError.first() }
```
Good call. I didn't make the entire method suspend (Timber documentation for planting/updating trees has historically been called from the main thread I believe, at least it's been that way since first implementation)... but I did move the code that checked for file access errors and called `runBlocking` on the dataStore to be in a new coroutine scope.

ssawchenko commented:
@phileo Hopefully I've addressed your comments. I've been playing around with DataStore a little more this afternoon and there's something I still don't like about having to pass in the Application/Context into the containing class, but I all of the examples I have found that create a `DataStoreRepository` type equivalent are doing it, so I guess that's the way forward for now.

phileo reviewed (approved):
re-tested, everything looks good! 👍
